File: sprint3/k/code.go
package main

import (
	"fmt"
	"reflect"
)

func merge_sort(arr []int, lf int, rg int) {
	if len(arr) <= 1 {
		return
	}
	mid := len(arr) / 2
	fmt.Println(mid)
	fmt.Println(arr)
	merge_sort(arr[:mid], lf, mid)
	merge_sort(arr[mid:], mid, rg)
	res := merge(arr, lf, mid, len(arr))
	for i := lf; i < len(arr); i++ {
		arr[i] = res[i-lf]
	}
}

func merge(arr []int, left int, mid int, right int) (result []int) {
	leftArr := arr[left:mid]
	rightArr := arr[mid:right]
	l, r := 0, 0
	for l < len(leftArr) && r < len(rightArr) {
		if leftArr[l] <= rightArr[r] {
			result = append(result, leftArr[l])
			l++
			continue
		}
		if leftArr[l] > rightArr[r] {
			result = append(result, rightArr[r])
			r++
			continue
		}
	}
	for l < len(leftArr) {
		result = append(result, leftArr[l])
		l++
	}
	for r < len(rightArr) {
		result = append(result, rightArr[r])
		r++
	}
	return
}

func main() {
	a := []int{1, 4, 9, 2, 10, 11}
	b := merge(a, 0, 3, 6)
	expected := []int{1, 2, 4, 9, 10, 11}
	if !reflect.DeepEqual(b, expected) {
		panic("WA. Merge")
	}

	c := []int{1, 4, 2, 10, 1, 2}
	merge_sort(c, 0, 6)
	fmt.Println(c)
	/*	expected = []int{1, 1, 2, 2, 4, 10}
		if !reflect.DeepEqual(c, expected) {
			panic("WA. MergeSort")
		}*/
}

File: sprint3/k/code.py
def merge(arr: list, left: int, mid: int, right: int) -> list:
	res = []
	left = arr[left:mid]
	right = arr[mid:right]
	mod_left = i_right = 0
	while len(res) < len(left) + len(right):
		if left[mod_left] <= right[i_right]:
			res.append(left[mod_left])
			mod_left += 1
		else:
			res.append(right[i_right])
			i_right += 1
		if i_right == len(right):
			res += left[mod_left:]
			break
		if mod_left == len(left):
			res += right[i_right:]
			break
	return res


def merge_sort(arr: list, left: int, right: int) -> list:
	if right - left >= 2:
		mid = (left + right) // 2
		merge_sort(arr, left, mid)
		merge_sort(arr, mid, right)
		arr[left:right] = merge(arr, left, mid, right)


def test():
	a = [1, 4, 9, 2, 10, 11]
	b = merge(a, 0, 3, 6)
	expected = [1, 2, 4, 9, 10, 11]
	assert b == expected
	c = [1, 4, 2, 10, 1, 2]
	merge_sort(c, 0, 6)
	expected = [1, 1, 2, 2, 4, 10]
	assert c == expected


if __name__ == '__main__':
	test()